sql/parser/expr.go

```diff
@@ -241,72 +241,156 @@ func (NullVal) String() string {
 
 // QualifiedName is a base name and an optional indirection expression.
 type QualifiedName struct {
-	Base     Name
-	Indirect Indirection
+	Base       Name
+	Indirect   Indirection
+	tableName  bool
+	columnName bool
 }
 
 // StarExpr is a convenience variable that represents an unqualified "*".
 var StarExpr = &QualifiedName{Indirect: Indirection{unqualifiedStar}}
 
+// NormalizeTableName normalizes the qualified name to contain a database name
+// as prefix, returning an error if unable to do so or if the name is not a
+// valid table name (e.g. it contains an array indirection).
+func (n *QualifiedName) NormalizeTableName(database string) error {
+	if n == nil || n.Base == "" {
+		return fmt.Errorf("empty table name: %s", n)
+	}
+	// The qualified name should have the form "table", "database.table",
+	// "table@index" or "database.table@index".
+	if len(n.Indirect) == 0 {
+		if database == "" {
+			return fmt.Errorf("no database specified")
+		}
+		n.Indirect = append(n.Indirect, NameIndirection(n.Base))
+		n.Base = Name(database)
+		n.tableName = true
+		return nil
+	}
+	// The qualified name should now have the form "database.table" or
+	// "database.table@index".
+	if len(n.Indirect) > 2 {
+		return fmt.Errorf("invalid table name: %s", n)
+	}
+	switch n.Indirect[0].(type) {
+	case NameIndirection:
+		// Nothing to do.
+	case IndexIndirection:
+		// "table@index"
+		if database == "" {
+			return fmt.Errorf("no database specified")
+		}
+		n.Indirect = append(Indirection{NameIndirection(n.Base)}, n.Indirect...)
+		n.Base = Name(database)
+	}
+	if len(n.Indirect) == 2 {
+		if _, ok := n.Indirect[1].(IndexIndirection); !ok {
+			return fmt.Errorf("invalid table name: %s", n)
+		}
+	}
+	n.tableName = true
+	return nil
+}
+
+// NormalizeColumnName normalizes the qualified name to contain a table name as
+// prefix, returning an error if unable to do so or if the name is not a valid
+// column name (e.g. it contains too many indirections). If normalization
+// occurred, the modified qualified name will have n.Base == "" to indicate no
+// explicit table was specified.
+func (n *QualifiedName) NormalizeColumnName() error {
+	if n == nil {
+		return fmt.Errorf("empty column name: %s", n)
+	}
+	// The qualified name should have the form "column", "*",
+	// "column[array-indirection]", "table.column", "table.*" or
+	// "table.column[array-indirection]".
+	if len(n.Indirect) == 0 {
+		if n.Base == "" {
+			return fmt.Errorf("empty column name: %s", n)
+		}
+		n.Indirect = append(n.Indirect, NameIndirection(n.Base))
+		n.Base = ""
+		n.columnName = true
+		return nil
+	}
+	if len(n.Indirect) > 2 {
+		return fmt.Errorf("invalid column name: %s", n)
+	}
+	switch n.Indirect[0].(type) {
+	case NameIndirection, StarIndirection:
+		// Nothing to do.
+	case *ArrayIndirection:
+		// "column[array-indirection]"
+		n.Indirect = append(Indirection{NameIndirection(n.Base)}, n.Indirect...)
+		n.Base = ""
```

Comment:
could you add a comment explaining what's going on here? Is this properly handling the array indirection?